File: tests/python/test_wilders.py
"""
Python binding tests for WILDERS indicator.
These tests mirror the Rust unit tests to ensure Python bindings work correctly.
"""
import pytest
import numpy as np
import sys
from pathlib import Path

try:
    import my_project as ta_indicators
except ImportError:
    # If not in virtual environment, try to import from installed location
    try:
        import my_project as ta_indicators
    except ImportError:
        pytest.skip("Python module not built. Run 'maturin develop --features python' first", allow_module_level=True)

from test_utils import load_test_data, assert_close, EXPECTED_OUTPUTS
from rust_comparison import compare_with_rust


class TestWilders:
    @pytest.fixture(scope='class')
    def test_data(self):
        return load_test_data()
    
    def test_wilders_partial_params(self, test_data):
        """Test Wilders with partial parameters (None values) - mirrors check_wilders_partial_params"""
        close = test_data['close']
        
        # Test with default period (5)
        result = ta_indicators.wilders(close, 5)
        assert len(result) == len(close)
    
    def test_wilders_accuracy(self, test_data):
        """Test Wilders matches expected values from Rust tests - mirrors check_wilders_accuracy"""
        close = test_data['close']
        expected = EXPECTED_OUTPUTS['wilders']
        
        result = ta_indicators.wilders(
            close,
            period=expected['default_params']['period']
        )
        
        assert len(result) == len(close)
        
        # Check last 5 values match expected
        assert_close(
            result[-5:], 
            expected['last_5_values'],
            rtol=1e-8,
            msg="Wilders last 5 values mismatch"
        )
        
        # Compare full output with Rust
        compare_with_rust('wilders', result, 'close', expected['default_params'])
    
    def test_wilders_default_candles(self, test_data):
        """Test Wilders with default parameters - mirrors check_wilders_default_candles"""
        close = test_data['close']
        
        # Default period: 5
        result = ta_indicators.wilders(close, 5)
        assert len(result) == len(close)
    
    def test_wilders_zero_period(self):
        """Test Wilders fails with zero period - mirrors check_wilders_zero_period"""
        input_data = np.array([10.0, 20.0, 30.0])
        
        with pytest.raises(ValueError, match="Invalid period"):
            ta_indicators.wilders(input_data, period=0)
    
    def test_wilders_period_exceeds_length(self):
        """Test Wilders fails when period exceeds data length - mirrors check_wilders_period_exceeds_length"""
        data_small = np.array([10.0, 20.0, 30.0])
        
        with pytest.raises(ValueError, match="Invalid period"):
            ta_indicators.wilders(data_small, period=10)
    
    def test_wilders_very_small_dataset(self):
        """Test Wilders with very small dataset - mirrors check_wilders_very_small_dataset"""
        single_point = np.array([42.0])
        
        # Should work with period=1
        result = ta_indicators.wilders(single_point, period=1)
        assert len(result) == 1
    
    def test_wilders_reinput(self, test_data):
        """Test Wilders applied twice (re-input) - mirrors check_wilders_reinput"""
        close = test_data['close']
        
        # First pass
        first_result = ta_indicators.wilders(close, period=5)
        assert len(first_result) == len(close)
        
        # Second pass - apply Wilders to Wilders output with different period
        second_result = ta_indicators.wilders(first_result, period=10)
        assert len(second_result) == len(first_result)
    
    def test_wilders_nan_handling(self, test_data):
        """Test Wilders handles NaN values correctly - mirrors check_wilders_nan_handling"""
        close = test_data['close']
        
        result = ta_indicators.wilders(close, period=5)
        assert len(result) == len(close)
        
        # After warmup period (240), no NaN values should exist
        if len(result) > 240:
            assert not np.any(np.isnan(result[240:])), "Found unexpected NaN after warmup period"
        
        # First period-1 values should be NaN (accounting for potential leading NaNs in data)
        # Find first non-NaN in input
        first_valid = np.argmax(~np.isnan(close))
        warmup_end = first_valid + 5 - 1  # period - 1
        assert np.all(np.isnan(result[:warmup_end])), "Expected NaN in warmup period"
    
    def test_wilders_streaming(self, test_data):
        """Test Wilders streaming matches batch calculation - mirrors check_wilders_streaming"""
        close = test_data['close']
        period = 5
        
        # Batch calculation
        batch_result = ta_indicators.wilders(close, period=period)
        
        # Streaming calculation
        stream = ta_indicators.WildersStream(period=period)
        stream_values = []
        
        for price in close:
            result = stream.update(price)
            stream_values.append(result if result is not None else np.nan)
        
        stream_values = np.array(stream_values)
        
        # Compare batch vs streaming
        assert len(batch_result) == len(stream_values)
        
        # Compare values where both are not NaN
        for i, (b, s) in enumerate(zip(batch_result, stream_values)):
            if np.isnan(b) and np.isnan(s):
                continue
            assert_close(b, s, rtol=1e-9, atol=1e-9, 
                        msg=f"Wilders streaming mismatch at index {i}")
    
    def test_wilders_batch(self, test_data):
        """Test Wilders batch processing - mirrors check_batch_default_row"""
        close = test_data['close']
        
        result = ta_indicators.wilders_batch(
            close,
            period_range=(5, 5, 0),  # Default period only
        )
        
        assert 'values' in result
        assert 'periods' in result
        
        # Should have 1 combination (default params)
        assert result['values'].shape[0] == 1
        assert result['values'].shape[1] == len(close)
        
        # Extract the single row
        default_row = result['values'][0]
        expected = EXPECTED_OUTPUTS['wilders']['last_5_values']
        
        # Check last 5 values match
        assert_close(
            default_row[-5:],
            expected,
            rtol=1e-8,
            msg="Wilders batch default row mismatch"
        )
    
    def test_wilders_all_nan_input(self):
        """Test Wilders with all NaN values"""
        all_nan = np.full(100, np.nan)
        
        with pytest.raises(ValueError, match="All values are NaN"):
            ta_indicators.wilders(all_nan, period=5)
    
    def test_wilders_kernel_selection(self, test_data):
        """Test Wilders with different kernel selections"""
        close = test_data['close']
        period = 5
        
        # Test different kernels
        for kernel in ['auto', 'scalar', 'avx2', 'avx512']:
            try:
                result = ta_indicators.wilders(close, period=period, kernel=kernel)
                assert len(result) == len(close)
            except ValueError as e:
                # AVX kernels might not be available on all systems
                if "Unknown kernel" not in str(e):
                    raise
    
    def test_wilders_batch_multiple_periods(self, test_data):
        """Test Wilders batch with multiple periods"""
        close = test_data['close'][:100]  # Use smaller dataset for speed
        
        result = ta_indicators.wilders_batch(
            close,
            period_range=(5, 10, 1),  # periods 5, 6, 7, 8, 9, 10
        )
        
        # Should have 6 rows (one for each period)
        assert result['values'].shape[0] == 6
        assert result['values'].shape[1] == len(close)
        
        # Verify periods array
        assert np.array_equal(result['periods'], [5, 6, 7, 8, 9, 10])
        
        # Verify each row has correct warmup
        for i, period in enumerate([5, 6, 7, 8, 9, 10]):
            row = result['values'][i]
            # First period-1 values should be NaN
            assert np.all(np.isnan(row[:period-1])), f"Expected NaN in warmup for period {period}"
            # After warmup should have values
            assert not np.any(np.isnan(row[period-1:])), f"Unexpected NaN after warmup for period {period}"


if __name__ == '__main__':
    pytest.main([__file__, '-v'])

File: tests/wasm/test_wilders.js
/**
 * WASM binding tests for WILDERS indicator.
 * These tests mirror the Rust unit tests to ensure WASM bindings work correctly.
 */
import test from 'node:test';
import assert from 'node:assert';
import path from 'path';
import { fileURLToPath } from 'url';
import { 
    loadTestData, 
    assertArrayClose, 
    assertClose,
    isNaN,
    assertAllNaN,
    assertNoNaN,
    EXPECTED_OUTPUTS 
} from './test_utils.js';
import { compareWithRust } from './rust-comparison.js';

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);

let wasm;
let testData;

test.before(async () => {
    
    try {
        const wasmPath = path.join(__dirname, '../../pkg/my_project.js');
        const importPath = process.platform === 'win32' 
            ? 'file:///' + wasmPath.replace(/\\/g, '/')
            : wasmPath;
        wasm = await import(importPath);
        
    } catch (error) {
        console.error('Failed to load WASM module. Run "wasm-pack build --features wasm --target nodejs" first');
        throw error;
    }
    
    testData = loadTestData();
});

test('Wilders partial params', () => {
    
    const close = new Float64Array(testData.close);
    
    const result = wasm.wilders_js(close, 5);
    assert.strictEqual(result.length, close.length);
});

test('Wilders accuracy', async () => {
    
    const close = new Float64Array(testData.close);
    const expected = EXPECTED_OUTPUTS.wilders;
    
    const result = wasm.wilders_js(
        close,
        expected.defaultParams.period
    );
    
    assert.strictEqual(result.length, close.length);
    
    
    const last5 = result.slice(-5);
    assertArrayClose(
        last5,
        expected.last5Values,
        1e-8,
        "Wilders last 5 values mismatch"
    );
    
    
    await compareWithRust('wilders', result, 'close', expected.defaultParams);
});

test('Wilders default candles', () => {
    
    const close = new Float64Array(testData.close);
    
    const result = wasm.wilders_js(close, 5);
    assert.strictEqual(result.length, close.length);
});

test('Wilders zero period', () => {
    
    const inputData = new Float64Array([10.0, 20.0, 30.0]);
    
    assert.throws(() => {
        wasm.wilders_js(inputData, 0);
    }, /Invalid period/);
});

test('Wilders period exceeds length', () => {
    
    const dataSmall = new Float64Array([10.0, 20.0, 30.0]);
    
    assert.throws(() => {
        wasm.wilders_js(dataSmall, 10);
    }, /Invalid period/);
});

test('Wilders very small dataset', () => {
    
    const singlePoint = new Float64Array([42.0]);
    
    
    const result = wasm.wilders_js(singlePoint, 1);
    assert.strictEqual(result.length, 1);
});

test('Wilders empty input', () => {
    
    const empty = new Float64Array([]);
    
    assert.throws(() => {
        wasm.wilders_js(empty, 5);
    }, /Input data slice is empty/);
});

test('Wilders NaN handling', () => {
    
    const close = new Float64Array(testData.close);
    
    const result = wasm.wilders_js(close, 5);
    assert.strictEqual(result.length, close.length);
    
    
    if (result.length > 240) {
        for (let i = 240; i < result.length; i++) {
            assert(!isNaN(result[i]), `Found unexpected NaN at index ${i}`);
        }
    }
    
    
    const warmupEnd = 4;
    assertAllNaN(result.slice(0, warmupEnd), "Expected NaN in warmup period");
    
    
    assert(!isNaN(result[warmupEnd]), `Expected valid value at index ${warmupEnd} (first output)`);
});

test('Wilders all NaN input', () => {
    
    const allNaN = new Float64Array(100);
    allNaN.fill(NaN);
    
    assert.throws(() => {
        wasm.wilders_js(allNaN, 5);
    }, /All values are NaN/);
});

test('Wilders batch single parameter set', () => {
    
    const close = new Float64Array(testData.close);
    
    
    const batchResult = wasm.wilders_batch_js(
        close,
        5, 5, 0      
    );
    
    
    const singleResult = wasm.wilders_js(close, 5);
    
    assert.strictEqual(batchResult.length, singleResult.length);
    assertArrayClose(batchResult, singleResult, 1e-10, "Batch vs single mismatch");
});

test('Wilders batch multiple periods', () => {
    
    const close = new Float64Array(testData.close.slice(0, 100)); 
    
    
    const batchResult = wasm.wilders_batch_js(
        close,
        5, 10, 1      
    );
    
    
    assert.strictEqual(batchResult.length, 6 * 100);
    
    
    const periods = [5, 6, 7, 8, 9, 10];
    for (let i = 0; i < periods.length; i++) {
        const rowStart = i * 100;
        const rowEnd = rowStart + 100;
        const rowData = batchResult.slice(rowStart, rowEnd);
        
        const singleResult = wasm.wilders_js(close, periods[i]);
        assertArrayClose(
            rowData, 
            singleResult, 
            1e-10, 
            `Period ${periods[i]} mismatch`
        );
    }
});

test('Wilders batch metadata', () => {
    
    const metadata = wasm.wilders_batch_metadata_js(
        5, 10, 1      
    );
    
    
    assert.strictEqual(metadata.length, 6);
    
    
    for (let i = 0; i < 6; i++) {
        assert.strictEqual(metadata[i], 5 + i);
    }
});

test('Wilders batch full parameter sweep', () => {
    
    const close = new Float64Array(testData.close.slice(0, 50));
    
    const batchResult = wasm.wilders_batch_js(
        close,
        5, 7, 2      
    );
    
    const metadata = wasm.wilders_batch_metadata_js(
        5, 7, 2
    );
    
    
    assert.strictEqual(metadata.length, 2);
    assert.strictEqual(batchResult.length, 2 * 50);
    
    
    for (let combo = 0; combo < 2; combo++) {
        const period = metadata[combo];
        const rowStart = combo * 50;
        const rowData = batchResult.slice(rowStart, rowStart + 50);
        const warmupEnd = period - 1;
        
        
        for (let i = 0; i < warmupEnd; i++) {
            assert(isNaN(rowData[i]), `Expected NaN at warmup index ${i} for period ${period}`);
        }
        
        
        assert(!isNaN(rowData[warmupEnd]), `Expected valid value at index ${warmupEnd} for period ${period}`);
        
        
        for (let i = warmupEnd; i < 50; i++) {
            assert(!isNaN(rowData[i]), `Unexpected NaN at index ${i} for period ${period}`);
        }
    }
});

test('Wilders batch unified API', () => {
    
    const close = new Float64Array(testData.close.slice(0, 100));
    
    
    if (typeof wasm.wilders_batch === 'function') {
        const result = wasm.wilders_batch(close, {
            period_range: [5, 7, 1]  
        });
        
        
        assert(result.values, 'Should have values array');
        assert(result.combos, 'Should have combos array');
        assert(typeof result.rows === 'number', 'Should have rows count');
        assert(typeof result.cols === 'number', 'Should have cols count');
        
        
        assert.strictEqual(result.rows, 3);
        assert.strictEqual(result.cols, 100);
        assert.strictEqual(result.combos.length, 3);
        assert.strictEqual(result.values.length, 300);
        
        
        for (let i = 0; i < 3; i++) {
            assert.strictEqual(result.combos[i].period, 5 + i);
        }
        
        
        const oldResult = wasm.wilders_js(close, 5);
        const firstRow = result.values.slice(0, 100);
        for (let i = 0; i < oldResult.length; i++) {
            if (isNaN(oldResult[i]) && isNaN(firstRow[i])) {
                continue; 
            }
            assert(Math.abs(oldResult[i] - firstRow[i]) < 1e-10,
                   `Value mismatch at index ${i}`);
        }
    } else {
        
        const batchResult = wasm.wilders_batch_js(close, 5, 7, 1);
        assert.strictEqual(batchResult.length, 3 * 100);
    }
});

test('Wilders batch edge cases', () => {
    
    const close = new Float64Array([1, 2, 3, 4, 5, 6, 7, 8, 9, 10]);
    
    
    const singleBatch = wasm.wilders_batch_js(
        close,
        5, 5, 1
    );
    
    assert.strictEqual(singleBatch.length, 10);
    
    
    const largeBatch = wasm.wilders_batch_js(
        close,
        5, 7, 10 
    );
    
    
    assert.strictEqual(largeBatch.length, 10);
    
    
    assert.throws(() => {
        wasm.wilders_batch_js(
            new Float64Array([]),
            5, 5, 0
        );
    }, /All values are NaN/);
});


test('Wilders zero-copy API', () => {
    const data = new Float64Array([1, 2, 3, 4, 5, 6, 7, 8, 9, 10]);
    const period = 5;
    
    
    const ptr = wasm.wilders_alloc(data.length);
    assert(ptr !== 0, 'Failed to allocate memory');
    
    
    const memory = wasm.__wasm.memory.buffer;
    const memView = new Float64Array(
        memory,
        ptr,
        data.length
    );
    
    
    memView.set(data);
    
    
    try {
        wasm.wilders_into(ptr, ptr, data.length, period);
        
        
        const regularResult = wasm.wilders_js(data, period);
        for (let i = 0; i < data.length; i++) {
            if (isNaN(regularResult[i]) && isNaN(memView[i])) {
                continue; 
            }
            assert(Math.abs(regularResult[i] - memView[i]) < 1e-10,
                   `Zero-copy mismatch at index ${i}: regular=${regularResult[i]}, zerocopy=${memView[i]}`);
        }
    } finally {
        
        wasm.wilders_free(ptr, data.length);
    }
});

test('Wilders zero-copy with large dataset', () => {
    const size = 10000;
    const data = new Float64Array(size);
    for (let i = 0; i < size; i++) {
        data[i] = Math.sin(i * 0.01) + Math.random() * 0.1;
    }
    
    const ptr = wasm.wilders_alloc(size);
    assert(ptr !== 0, 'Failed to allocate large buffer');
    
    try {
        const memory = wasm.__wasm.memory.buffer;
        const memView = new Float64Array(memory, ptr, size);
        memView.set(data);
        
        wasm.wilders_into(ptr, ptr, size, 5);
        
        
        const memory2 = wasm.__wasm.memory.buffer;
        const memView2 = new Float64Array(memory2, ptr, size);
        
        
        for (let i = 0; i < 4; i++) {
            assert(isNaN(memView2[i]), `Expected NaN at warmup index ${i}`);
        }
        
        
        for (let i = 4; i < Math.min(100, size); i++) {
            assert(!isNaN(memView2[i]), `Unexpected NaN at index ${i}`);
        }
    } finally {
        wasm.wilders_free(ptr, size);
    }
});

test('Wilders zero-copy error handling', () => {
    
    assert.throws(() => {
        wasm.wilders_into(0, 0, 10, 5);
    }, /null pointer|invalid memory/i);
    
    
    const ptr = wasm.wilders_alloc(10);
    try {
        
        assert.throws(() => {
            wasm.wilders_into(ptr, ptr, 10, 0);
        }, /Invalid period/);
        
        
        assert.throws(() => {
            wasm.wilders_into(ptr, ptr, 10, 20);
        }, /Invalid period/);
    } finally {
        wasm.wilders_free(ptr, 10);
    }
});

test('Wilders memory management', () => {
    
    const sizes = [100, 1000, 10000];
    
    for (const size of sizes) {
        const ptr = wasm.wilders_alloc(size);
        assert(ptr !== 0, `Failed to allocate ${size} elements`);
        
        
        const memory = wasm.__wasm.memory.buffer;
        const memView = new Float64Array(memory, ptr, size);
        for (let i = 0; i < Math.min(10, size); i++) {
            memView[i] = i * 1.5;
        }
        
        
        for (let i = 0; i < Math.min(10, size); i++) {
            assert.strictEqual(memView[i], i * 1.5, `Memory corruption at index ${i}`);
        }
        
        
        wasm.wilders_free(ptr, size);
    }
});

test('Wilders NaN in initial window', () => {
    
    const data = new Float64Array([1.0, 2.0, NaN, 4.0, 5.0, 6.0, 7.0, 8.0, 9.0, 10.0]);
    
    assert.throws(() => {
        wasm.wilders_js(data, 5);
    }, /Not enough valid data/);
});

test.after(() => {
    console.log('Wilders WASM tests completed');
});
